core/src/env/engine/off_chain/chain_extension.rs

```diff
@@ -0,0 +1,112 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::OffChainError;
+use crate::env::Result;
+use std::collections::HashMap;
+
+type FuncId = u32;
+
+/// Types implementing this trait can be used as chain extensions.
+///
+/// This trait is only useful for testing contract via the off-chain environment.
+pub trait ChainExtension {
+    /// The expected input type.
+    ///
+    /// # Note
+    ///
+    /// This can be a tuple to expect multiple input types.
+    type Input: scale::Codec;
+    /// The expected output type.
+    type Output: scale::Codec;
+
+    /// The static function ID of the chain extension.
+    ///
+    /// # Note
+    ///
+    /// This is expected to return a constant value.
+    fn func_id(&self) -> u32;
+
+    /// Calls the chain extension with the given input.
+    fn call(&mut self, input: &Self::Input) -> Result<Self::Output>;
+}
+
+/// A raw chain extension function.
+///
+/// This is mostly a wrapper closure around the real chain extension function
+/// that handles marshalling of types between their encoded and decoded
+/// representations.
+type ChainExtensionFn = Box<dyn FnMut(Vec<u8>) -> Result<Vec<u8>>>;
+
+/// Runtime call handler.
+///
+/// More generically a mapping from bytes to bytes.
+pub struct ChainExtensionHandler {
+    /// The currently registered runtime call handler.
+    registered: HashMap<FuncId, ChainExtensionFn>,
+}
+
+impl ChainExtensionHandler {
+    /// Creates a new chain extension handler.
+    ///
+    /// Initialized with an empty set of chain extensions.
+    pub fn new() -> Self {
+        Self {
+            registered: HashMap::new(),
+        }
+    }
+
+    /// Resets the chain extension handler to uninitialized state.
+    pub fn reset(&mut self) {
+        self.registered.clear()
+    }
+
+    /// Register a new chain extension.
+    pub fn register<I, O>(
+        &mut self,
+        mut extension: Box<dyn ChainExtension<Input = I, Output = O>>,
+    ) where
+        I: scale::Codec + 'static,
+        O: scale::Codec + 'static,
+    {
+        let func_id = extension.func_id();
+        self.registered.insert(
+            func_id,
+            Box::new(move |encoded_input: Vec<u8>| {
+                let decoded_input = scale::Decode::decode(&mut &encoded_input[..])?;
+                let decoded_output = extension.call(&decoded_input)?;
+                Ok(scale::Encode::encode(&decoded_output))
+            }),
+        );
+    }
+
+    /// Invokes the chain extension with the given parameters.
+    pub fn invoke<I, O>(&mut self, func_id: FuncId, input: &I) -> Result<O>
```

Comment:
AFAIK ink! uses "call" as verb in other places when a result is returned and not invoke.

Reply:
This might be a good idea. The problem I have with `call` is that it doesn't tell the user anything about any evaluation of return type which to my understanding is better expressed by invoke/eval pairs.

Reply:
So it should be called eval then?

Reply:
That's actually a fair point! :)